src/scripts/bttvIntegration.js

```diff
@@ -0,0 +1,137 @@
+ * @returns {string} the base text modified with any matched keywords turned into html image tags with that emote's as src
+ */
+function replaceKeywordWithBttvEmoteImage(textContent, keyword, imageSource, id){ 
+    const argumentsExistAndAreStrings = areAllTruthyStrings([textContent,keyword,imageSource,id]); 
+    if (!argumentsExistAndAreStrings) { return null }
+    const imageHtmlString = 
+    `<img alt="${keyword}" data-twitch-emote="${keyword}" data-twitch-emote-id="${id}" data-twitch-emote-source="bttv" src="${imageSource}"></img>`;
+    const words = textContent.split(' ');
+    const modifiedTextContent = words.map(word => word === keyword ? imageHtmlString : word).join(' ');
+    return modifiedTextContent;
+}
+
+
+/**
+ * @param {Array} arrayOfInputs values to be tested 
+ * @returns {Boolean} true if all inputs provided are both truthy and strings, false if not
+ */
+function areAllTruthyStrings(arrayOfInputs)  { 
+    let areTruthyStrings = true;
+    if (arrayOfInputs.length === 0){ areTruthyStrings = false }
+    for (input of arrayOfInputs){ 
+        if (!input || typeof input !== "string"){ areTruthyStrings = false }
+    }
+    return areTruthyStrings
+}
+
+
+/**
+ * @param {string} username twitch username, case insensitive
+ * @returns {string} twitch user id
+ */
+async function getTwitchUserId(username, fetchFunction){ 
+    if (!username || typeof username !== "string"){ 
+        return null;
+    }
+    fetchFunction = fetchFunction || defaultFetch
+    const apiBaseUrl = "https://streamraiders.tips/_functions/getTwitchProfileData/";
+    const urlToQuery = apiBaseUrl + username;
+    const twitchUserData = await fetchFunction(urlToQuery);
+    return new Promise( resolve => { 
+        resolve(twitchUserData.id)
+    })
+}
+
+
+/**
+ * @param {string} userId twitch's user ID 
+ * @param {Function} fetchFunction OPTIONAL - a specific fetch function that returns a promise
+ * that resolves to a json of the fetch data.
+ * @returns {Object} key is the friendly emote name, value is the BTTV emote id
+ */
+async function getBttvChannelEmoteDict(userId, fetchFunction){
+    if ( !userId || typeof userId !== "string") {return null} 
+    fetchFunction = fetchFunction || defaultFetch
+    const apiBaseUrl = "https://api.betterttv.net/3/cached/users/twitch/"
+    const urlToQuery = apiBaseUrl + userId;
+    const bttvChannelData = await fetchFunction(urlToQuery);
+    const emoteDict = convertBttvChannelDataToEmoteDict(bttvChannelData);
+    return new Promise( resolve=> { 
+        resolve(emoteDict);
+    })
+}
+
+
+/**
+ * @param {string} url url to retrieve json from
+ * @returns {Object} json data received from fetch
+ */
+async function defaultFetch(url){ 
+    const response = await fetch(url);
+    const json = await response.json();
+    return new Promise(resolve => { 
+        resolve(json);
+    })
+}
+
+
+/**
+ * @param {Object} channelData  data structure provided by BTTV's api for channel queries
```

Comment:
Do we have a sense of what properties to expect? If so, we can call those out more specifically here. We've got a couple ways to do that:

One is inline:

```js
/**
 * @param {{
 *    channelEmotes: {code: string, id: string}[],
 *    sharedEmotes: {code: string, id: string}[]
 * }} channelData data structure provided by BTTV's api for channel queries
 */
```

You can also use JSDoc's `@typedef` feature to extract inline types into their own named declaration:

```js
/**
 * typedef {Object} BttvEmote
 * @property {string} code emote name (substring to replace from chat messages)
 * @property {id} unique identifier for emote (used for forming the emote image source)
 */

/**
 * @param {{
 *    channelEmotes: BttvEmote[],
 *    sharedEmotes: BttvEmote[]
 * }} channelData data structure provided by BTTV's api for channel queries
 */
```

And if it makes sense, especially if we get a *lot* more than this in BTTV's response, we can do more extraction:

```js
/**
 * typedef {Object} BttvEmote
 * @property {string} code emote name (substring to replace from chat messages)
 * @property {id} unique identifier for emote (used for forming the emote image source)
 */

/**
 * @typedef {Object} BttvChannelData
 * @property {BttvEmote[]} channelEmotes
 * @property {BttvEmote[]} sharedEmotes
 */

/**
 * @param {BttvChannelData} channelData data structure provided by BTTV's api for channel queries
 */
```

I haven't seen the full response from the API, so going full type extraction might be overkill, but are there ways we could implement some of this to add a bit more specificity to our JSDoc?

Reply:
FWIW, this definitely isn't a "you have to do all of this" — just providing possibilities and leaving it open for your input!